components/tikv_util/src/quota_limiter.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::config::ReadableSize;
+use super::time::Limiter;
+use cpu_time::ThreadTime;
+use std::time::Duration;
+use std::time::Instant;
+
+const CPU_TIME_FACTOR: f64 = 0.9;
+
+// Quota limiter allows users to obtain stable performance by increasing the
+// completion time of tasks through restrictions of different metrics.
+pub struct QuotaLimiter {
+    cputime_limiter: Limiter,
+    write_kvs_limiter: Limiter,
+    write_bandwidth_limiter: Limiter,
+    read_bandwidth_limiter: Limiter,
+    req_rate_limiter: Limiter,
+}
+
+impl Default for QuotaLimiter {
+    fn default() -> Self {
+        Self {
+            cputime_limiter: Limiter::new(f64::INFINITY),
+            write_kvs_limiter: Limiter::new(f64::INFINITY),
+            write_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            read_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            req_rate_limiter: Limiter::new(f64::INFINITY),
+        }
+    }
+}
+
+impl QuotaLimiter {
+    // 1000 millicpu equals to 1vCPU, 0 means unlimited
+    pub fn new(
+        cpu_quota: usize,
+        req_rate: usize,
+        write_kvs: usize,
+        write_bandwidth: ReadableSize,
+        read_bandwidth: ReadableSize,
+    ) -> Self {
+        let cputime_limiter = if cpu_quota == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            // The cpu time is not a real statistics, only part of the processing logic is
+            // taken into account, so it needs to be multiplied by a factor.
+            // transfer milli cpu to micro cpu
+            Limiter::new(cpu_quota as f64 * CPU_TIME_FACTOR * 1000_f64)
+        };
+
+        let write_kvs_limiter = if write_kvs == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_kvs as f64)
+        };
+
+        let write_bandwidth_limiter = if write_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_bandwidth.0 as f64)
+        };
+
+        let read_bandwidth_limiter = if read_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(read_bandwidth.0 as f64)
+        };
+
+        let total_qps_limiter = if req_rate == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(req_rate as f64)
+        };
+
+        Self {
+            cputime_limiter,
+            write_kvs_limiter,
+            write_bandwidth_limiter,
+            read_bandwidth_limiter,
+            req_rate_limiter: total_qps_limiter,
+        }
+    }
+
+    pub fn consume_write(
+        &self,
+        req_cnt: usize,
+        kv_cnt: usize,
+        bytes: usize,
+        cpu_micro_secs: usize,
+    ) -> Duration {
+        let cpu_dur = self.cputime_limiter.consume_duration(cpu_micro_secs);
```

Comment:
add check to totally skip consume_duration for better performance when cpu limiter is infinity.

same as other limiters.

Reply:
explanation in https://github.com/tikv/tikv/pull/12151#issuecomment-1066823807